File: PROJECTS/CHALLENGE_RR/find_trajectory.m
% FIND_TRAJECTORY: find resulting trajectory inbetween two points
% SANTIAGO GARCIA ARANGO


function [x, y] = find_trajectory(trajectory_type, x_inicial, x_final,...
                                    y_inicial, y_final, time)

    if trajectory_type == "linear"
        % Find line equation of the robot
        m = (y_final - y_inicial)/(x_final - x_inicial);
        b = y_final - m*x_final;

        % Generate the vectors for each point of the robot movement
        x = linspace(x_inicial, x_final, time);
        y = m*x + b;
    end

    
    if trajectory_type == "5th"
        % Use input parameters for the variables needed
        x_o = x_inicial;
        y_o = y_inicial;
        x_f = x_final;
        y_f = y_final;

        % Define the intitial conditions for the fifth order trajectory    
        x_d_o = 0;
        y_d_o = 0;
        x_d_f = 0;
        y_d_f = 0;
        x_dd_o = 0;
        y_dd_o = 0;
        x_dd_f = 0;
        y_dd_f = 0;
        
        tf = time;
        
        % Apply constants for the fifth order trajectories (see Craig book)
        a0_x = x_o;
        a1_x = x_d_o;
        a2_x = x_dd_o/2;
        a3_x = (20*x_f - 20*x_o - (8*x_d_f + 12*x_d_o)*tf - (3*x_dd_o - x_dd_f)*tf^2)/(2*tf^3);
        a4_x = (30*x_o - 30*x_f + (14*x_d_f + 16*x_d_o)*tf + (3*x_dd_o - 2*x_dd_f)*tf^2)/(2*tf^4);
        a5_x = (12*x_f - 12*x_o - (6*x_d_f + 6*x_d_o)*tf - (x_dd_o - x_dd_f)*tf^2)/(2*tf^5);
        
        a0_y = y_o;
        a1_y = y_d_o;
        a2_y = y_dd_o/2;
        a3_y = (20*y_f - 20*y_o - (8*y_d_f + 12*y_d_o)*tf - (3*y_dd_o - y_dd_f)*tf^2)/(2*tf^3);
        a4_y = (30*y_o - 30*y_f + (14*y_d_f + 16*y_d_o)*tf + (3*y_dd_o - 2*y_dd_f)*tf^2)/(2*tf^4);
        a5_y = (12*y_f - 12*y_o - (6*y_d_f + 6*y_d_o)*tf - (y_dd_o - y_dd_f)*tf^2)/(2*tf^5);
    
        % Generate the vectors for each point of the robot movement
        t = linspace(x_o, x_f, time);
        x = zeros(1, size(t, 2));
        y = zeros(1, size(t, 2));
        
        
    end

    
    
end

File: PROJECTS/CHALLENGE_BAXTER_PY/find_trajectory.py
#!/usr/bin/env python
# FIND 5TH ORDER TRAJECTORY INBETWEEN INITIAL AND FINAL POINT
# SANTIAGO GARCIA AND ELKIN GUERRA

import numpy as np
import math

def find_trajectory(var_initial, var_final, final_time):

    # Use input parameters for the variables needed
    var_o = var_initial
    var_f = var_final

    # Define the intitial conditions for the fifth order trajectory    
    var_d_o = 0
    var_d_f = 0
    var_dd_o = 0
    var_dd_f = 0

    tf = final_time

    # Apply constants for the fifth order trajectories (Robotic's Craig book)
    a0 = var_o
    a1 = var_d_o
    a2 = var_dd_o/2
    a3 = (20*var_f - 20*var_o - (8*var_d_f + 12*var_d_o)*tf - (3*var_dd_o - var_dd_f)*tf**2)/(2*tf**3)
    a4 = (30*var_o - 30*var_f + (14*var_d_f + 16*var_d_o)*tf + (3*var_dd_o - 2*var_dd_f)*tf**2)/(2*tf**4)
    a5 = (12*var_f - 12*var_o - (6*var_d_f + 6*var_d_o)*tf - (var_dd_o - var_dd_f)*tf**2)/(2*tf**5)


    # Initialize the vectors for the important variables of the robot
    t = np.linspace(0, tf, 50)
    var = np.zeros(np.size(t))        

    # Generate main position vectors based on the constants found
    for i in range(len(t)):
        var[i] = a0 + a1*t[i] + a2*(t[i])**2 + a3*(t[i])**3 + a4*(t[i])**4 + a5*(t[i])**5

    return var


# TESTS
if __name__ == "__main__":
    theta1 = find_trajectory(0.5, 1.5, 5)
    print("SIZE(theta1) = {}".format(np.size(theta1)))
    print("theta1 = \n{}".format(theta1))
